src/Illuminate/Database/Eloquent/Model.php

```diff
@@ -3177,19 +3177,19 @@ public function syncOriginalAttribute($attribute)
      */
     public function isDirty($attributes = null)
     {
-        $dirty = $this->getDirty();
-
         if (is_null($attributes)) {
-            return count($dirty) > 0;
-        }
-
-        if (! is_array($attributes)) {
+            $attributes = array_keys($this->attributes);
+        } elseif (! is_array($attributes)) {
             $attributes = func_get_args();
         }
 
-        foreach ($attributes as $attribute) {
-            if (array_key_exists($attribute, $dirty)) {
-                return true;
+        foreach ($attributes as $key) {
+            if (isset($this->attributes[$key])) {
+                if (! array_key_exists($key, $this->original) ||
+                        $this->attributes[$key] !== $this->original[$key] &&
+                            ! $this->originalIsNumericallyEquivalent($key)) {
```

Comment:
This condition is quite hard to read. Is it intended that there is no logical grouping... You have `foo || bar && baz` with no grouping.